Add `DROP DATABASE name CASCADE ALIASES` (#1805)
neo4j/docs-operations: modules/ROOT/pages/database-administration/composite-databases/delete-composite-databases.adoc

For more information, see xref:database-administration/aliases/manage-aliases-composite-databases.adoc#delete-composite-database-alias[Delete database aliases in composite databases].

You can delete composite databases using either the command `DROP COMPOSITE DATABASE name` or the more general one `DROP DATABASE name`.
However, keep in mind that the first command targets only composite databases, while the second one targets any database.

.Query
[source, cypher]
----
DROP COMPOSITE DATABASE inventory
----

[role=enterprise-edition label--new-5.24]
[[composite-databases-delete-with-aliases]]
== Delete a composite database while dropping its constituents

You can use the `CASCADE ALIASES` option of the `DROP COMPOSITE DATABASE` Cypher command to drop the constituent database aliases while deleting the composite database.

[NOTE]
====
This operation does not deletes the actual target databases of the constituent database aliases.
====

The `CASCADE ALIASES` option is useful when you want to delete a composite database and its constituent database aliases in one step.
Using `CASCADE ALIASES` requires the `DROP ALIAS` privilege.

> For more information, see xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-alias-management[ALIAS MANAGEMENT privileges].

Author: ```suggestion
For more information about this privilege, see xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-alias-management[ALIAS MANAGEMENT privileges].
```
since that page don't really talk about drop database cascade

.Drop a composite database and its consitutent alias
====
In this example, we create a composite database `movies` and a database alias `movies.sweden` for the database `swedish-movies` and then delete the alias `sweden` and the composite database `movies`.

.Create a composite database `movies` and a database alias `movies.sweden` for the database `swedish-movies`
[source, cypher]
----
CREATE COMPOSITE DATABASE movies
CREATE ALIAS movies.sweden FOR DATABASE `swedish-movies`
----

.Delete the composite database `movies` while also dropping the alias `movies.sweden`
[source, cypher]
----
DROP COMPOSITE DATABASE movies CASCADE ALIASES
----
====

This behavior is the same for the more general command `DROP DATABASE name` when using it to drop a composite database.

> The option `RESTRICT` explicitly requests the default behavior of the command.

Author: since this section don't have a syntax description nor mention RESTRICT before this point, this sentence sticks out weirdly

Reviewer: Yes, I had this question in my mind: What is the default behavior of the command?

Author: to throw an error when there exists aliases (for composite its constituent aliases and for standard it's local ones targeting it), so the behaviour if you don't write `CASCADE ALIASES`

Reviewer: Ah ok. And when or why would one use it?

Author: I don't really see people using it as it means the same as leaving it out, but it is there for consistency. We do allow you to specify the default behaviours of DUMP DATA/NOWAIT so why wouldn't we allow you to specify RESTRICT

Author: I think leaving it out on the composite page is fine

> The option `RESTRICT` explicitly requests the default behavior of the command.

Reviewer: ```suggestion
//The option `RESTRICT` explicitly requests the default behavior of the command.
```

[[composite-databases-delete]]
= Delete composite databases

> You can delete composite databases using either the command `DROP COMPOSITE DATABASE name` or the more general one  `DROP DATABASE name`.

Reviewer 2: Not sure this should be 'You'. Probably this paragraph needs some editing.

Author: this is the current version (I just removed a space where there were 2 between the words), I'll leave it to the docs team to decide on the `you`

DROP COMPOSITE DATABASE inventory
----

[[composite-databases-delete-with-aliases]]
== Delete a composite database with constituent database aliases

In general, deleting a composite database that has constituent database aliases will fail due to those database aliases.
This affects both the local and remote constituent database aliases belonging to the specified composite database.

By appending `CASCADE ALIASES` to the command `DROP COMPOSITE DATABASE`, those constituent database aliases can be dropped together with the composite database.

> This affects both the local and remote constituent database aliases belonging to the specified composite database.

Reviewer 2: Do we need to repeat this here?

> This affects both the local and remote constituent database aliases belonging to the specified composite database.

Reviewer 2: Yeah, maybe we should mention that. I don't think you can overdocument cascading deletes.

DROP COMPOSITE DATABASE inventory
----

[[composite-databases-delete-with-aliases]]
== Delete a composite database with constituent database aliases

> In general, deleting a composite database that has constituent database aliases will fail due to those database aliases.

Reviewer 2: ```suggestion
In general, deleting a composite database that has constituent database aliases will fail as these aliases need to be dropped first.
```

> This affects both the local and remote constituent database aliases belonging to the specified composite database.

Reviewer 2: ```suggestion
This applies to both the local and remote constituent database aliases belonging to the specified composite database.
```

> By appending `CASCADE ALIASES` to the command `DROP COMPOSITE DATABASE`, those constituent database aliases can be dropped together with the composite database.

Reviewer 2: ```suggestion
By appending `CASCADE ALIASES` to the command `DROP COMPOSITE DATABASE`, those constituent database aliases can be dropped alongside the composite database.
```

This affects both the local and remote constituent database aliases belonging to the specified composite database.

Using `CASCADE ALIASES` requires the `DROP ALIAS` privilege (see xref:authentication-authorization/dbms-administration.adoc#access-control-dbms-administration-alias-management[ALIAS MANAGEMENT privileges]).

Given composite database and database alias:
[source, cypher]
----
CREATE COMPOSITE DATABASE movies
CREATE ALIAS movies.sweden FOR DATABASE `swedish-movies`
----

Then both the constituent database alias `sweden` and the composite database `movies` will be dropped with:
[source, cypher]
----
DROP COMPOSITE DATABASE movies CASCADE ALIASES
----

> This behaviour would be the same for the more general one `DROP DATABASE name` as well when using it to drop composite databases.

Reviewer 2: ```suggestion
This behaviour would be the same for the more general command `DROP DATABASE name` when using it to drop a composite database.
```
Also not sure about the British spelling behaviour, maybe it should be the US behavior

Author: it probably should yes
```suggestion
This behavior would be the same for the more general command `DROP DATABASE name` when using it to drop a composite database.
```

The option `RESTRICT` explicitly requests the default behavior of the command.

[NOTE]
====

> The `CASCADE ALIASES` option will drop the same aliases that would be given in the error message from trying to drop the database using the `RESTRICT` mode instead.

Reviewer 2: Do we need this note? It seems like it could easily get broken by changing the error message and it kinda restates the above.

Reviewer 2: I'm not entirely convinced by the comment, but if we want to address it, I don't think this is a good way because running the command to get the error message will actually drop the database if there are no aliases.

Perhaps something like: 

The aliases that will be dropped can be shown with `SHOW DATABASE db YIELD name, constituents` . And the same for aliases for the other page.

Author: I mean you were kinda after trying to drop the database regardless 🤷 but I'll update to something along those lines tomorrow 👍
